Unit test generators are unsupported

Due to ["limited support"](https://github.com/xmlrunner/unittest-xml-reporting#limited-support-for-unittesttestcasesubtest) in xmlrunner for unittest.TestCase.subTest and the fact that pybuilder uses xmlrunner to run unit tests, unit test generators are unsupported in pybuilder, which is a shame because they're a really nice feature :)

## Expected Behavior

Tests run - 3 pass, 3 fail.

## Actual Behavior

BUILD FAILED - TypeError: issubclass() arg 2 must be a class or tuple of classes (xmlrunner/result.py:375)

```testcase_pyb pyb -v
PyBuilder version 0.12.6
Build started at 2020-07-17 12:58:14
------------------------------------------------------------
[INFO]  Building testcase_pyb version 1.0.dev0
[INFO]  Executing build in /Users/neil/testcase_pyb
[INFO]  Going to execute task publish
[INFO]  Processing plugin packages 'coverage~=5.0' to be installed with {'upgrade': True}
[INFO]  Processing plugin packages 'flake8~=3.7' to be installed with {'upgrade': True}
[INFO]  Processing plugin packages 'pypandoc~=1.4' to be installed with {'upgrade': True}
[INFO]  Processing plugin packages 'setuptools>=38.6.0' to be installed with {'upgrade': True}
[INFO]  Processing plugin packages 'twine>=1.15.0' to be installed with {'upgrade': True}
[INFO]  Processing plugin packages 'unittest-xml-reporting~=2.5.2' to be installed with {'upgrade': True}
[INFO]  Processing plugin packages 'wheel>=0.34.0' to be installed with {'upgrade': True}
[INFO]  Creating target 'build' VEnv in '/Users/neil/testcase_pyb/target/venv/build/cpython-3.7.7.final.0'
[INFO]  Creating target 'test' VEnv in '/Users/neil/testcase_pyb/target/venv/test/cpython-3.7.7.final.0'
[INFO]  Requested coverage for tasks: pybuilder.plugins.python.unittest_plugin:run_unit_tests
[INFO]  Running unit tests
[INFO]  Executing unit tests from Python modules in /Users/neil/testcase_pyb/src/unittest/python
------------------------------------------------------------
BUILD FAILED - TypeError: issubclass() arg 2 must be a class or tuple of classes (xmlrunner/result.py:375)
------------------------------------------------------------
Build finished at 2020-07-17 12:58:59
Build took 44 seconds (44762 ms)```
```

## Steps to Reproduce the Problem

1.  Initialise a new pybuilder project
2. Write the following to ./src/unittest/python/generator_tests.py (The example code taken from [here](https://docs.python.org/3.4/library/unittest.html#distinguishing-test-iterations-using-subtests))

```
import unittest

class NumbersTest(unittest.TestCase):

    def test_even(self):
        """
        Test that numbers between 0 and 5 are all even.
        """
        for i in range(0, 6):
            with self.subTest(i=i):
                self.assertEqual(i % 2, 0)
```

3. Run the tests with pyb -v

## Specifications

  - Version: pyb 0.12.6, Python 3.7.7
  - Platform: MacOS Catalina (10.15.5)

--- a/src/main/python/pybuilder/plugins/python/coverage_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/coverage_plugin.py
@@ -41,7 +41,7 @@ use_plugin("analysis")
 
 @init
 def init_coverage_properties(project):
-    project.plugin_depends_on("coverage", "~=5.0")
+    project.plugin_depends_on("coverage", "~=5.2")
 
     # These settings are for aggregate coverage
     project.set_property_if_unset("coverage_threshold_warn", 70)

--- a/src/main/python/pybuilder/plugins/python/coveralls_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/coveralls_plugin.py
@@ -23,6 +23,7 @@ from pybuilder.core import init, use_plugin, finalize
 from pybuilder.errors import BuildFailedException
 from pybuilder.execution import ExecutionManager
 from pybuilder.plugins.python._coverage_util import patch_coverage
+from pybuilder.python_utils import PY2
 
 use_plugin("python.core")
 use_plugin("analysis")
@@ -31,7 +32,10 @@ use_plugin("python.coverage")
 
 @init(environments="ci")
 def init_coveralls_properties(project):
-    project.plugin_depends_on("coveralls", "~=1.11")
+    if PY2:
+        project.plugin_depends_on("coveralls", "<2.0.0")
+    else:
+        project.plugin_depends_on("coveralls", "~=2.1")
 
     project.set_property_if_unset("coveralls_dry_run", False)
     project.set_property_if_unset("coveralls_report", False)

--- a/src/main/python/pybuilder/plugins/python/unittest_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/unittest_plugin.py
@@ -37,9 +37,11 @@ use_plugin("python.core")
 
 @init
 def init_test_source_directory(project):
-    project.plugin_depends_on("unittest-xml-reporting", "~=2.5.2")
     if PY2:
         project.plugin_depends_on("mock")
+        project.plugin_depends_on("unittest-xml-reporting", "~=2.5.2")
+    else:
+        project.plugin_depends_on("unittest-xml-reporting", "~=3.0.2")
 
     project.set_property_if_unset("dir_source_unittest_python", "src/unittest/python")
     project.set_property_if_unset("unittest_breaks_build", True)

--- a/src/main/python/pybuilder/remote/__init__.py
+++ b/src/main/python/pybuilder/remote/__init__.py
@@ -23,6 +23,24 @@ patch_mp()
 
 import sys  # noqa: E402
 from textwrap import dedent  # noqa: E402
+
+from types import (MethodType,  # noqa: E402
+                   FunctionType,
+                   BuiltinFunctionType,
+                   )
+
+try:
+    from types import (WrapperDescriptorType,  # noqa: E402
+                       MethodWrapperType,
+                       MethodDescriptorType,
+                       ClassMethodDescriptorType,
+                       )
+except ImportError:
+    WrapperDescriptorType = type(object.__init__)
+    MethodWrapperType = type(object().__str__)
+    MethodDescriptorType = type(str.join)
+    ClassMethodDescriptorType = type(dict.__dict__['fromkeys'])
+
 from os.path import normcase as nc, sep  # noqa: E402
 from io import BytesIO, StringIO  # noqa: E402
 from pickle import PickleError, Unpickler, UnpicklingError, HIGHEST_PROTOCOL  # noqa: E402
@@ -36,6 +54,15 @@ from pybuilder.python_utils import (mp_get_context,  # noqa: E402
 
 PICKLE_PROTOCOL_MIN = 2
 PICKLE_PROTOCOL_MAX = HIGHEST_PROTOCOL
+
+CALLABLE_TYPES = (MethodType,
+                  FunctionType,
+                  BuiltinFunctionType,
+                  MethodWrapperType,
+                  ClassMethodDescriptorType,
+                  WrapperDescriptorType,
+                  MethodDescriptorType,
+                  )
 
 if PY2:
     ConnectionError = EnvironmentError
@@ -368,7 +395,7 @@ def proxy_members(obj, public=True, protected=True, add_callable=True, add_str=T
                 is_callable and add_callable or
                 is_str and add_str):
             member = getattr(obj, name)
-            if callable(member):
+            if isinstance(member, CALLABLE_TYPES):
                 methods.append(name)
             else:
                 fields.append(name)

--- a/src/unittest/python/plugins/python/unittest_plugin_tests.py
+++ b/src/unittest/python/plugins/python/unittest_plugin_tests.py
@@ -26,6 +26,7 @@ from pybuilder.plugins.python.unittest_plugin import (execute_tests, execute_tes
                                                       _create_runner,
                                                       _get_make_result_method_name,
                                                       report_to_ci_server)
+from pybuilder.python_utils import PY2
 from pybuilder.utils import np
 from test_utils import Mock, patch
 
@@ -212,3 +213,15 @@ class UnittestRunnerTest(TestCase):
             pass
 
         self.assertEqual(_get_make_result_method_name((TextTestRunner, _makeResult)), "_makeResult")
+
+
+class UnittestRunnerCompatibilityTest(TestCase):
+    def test_sub_tests_issue_735(self):
+        """
+        Test that numbers between 0 and 5 are all between 0 and 5.
+        """
+        if not PY2:
+            for i in range(0, 6):
+                with self.subTest(i=i):
+                    self.assertLess(i, 6)
+                    self.assertGreaterEqual(i, 0)

